In [1]:
%%bash
cd /core
git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*" && git fetch && git checkout main
pip install . --no-deps

cd /dataloader
git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*" && git fetch && gheckout main
pip install . --no-deps

cd /nvtabular
git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*" && git fetch && git checkout main
pip install . --no-deps

cd /workspace
pip install . --no-deps

cd /systems
git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*" && git fetch && git checkout main
pip install . --no-deps

cd /transformers4rec
git config remote.origin.fetch "+refs/heads/*:refs/remotes/origin/*" && git fetch && git checkout main
pip install . --no-deps

From https://github.com/NVIDIA-Merlin/core
   9d9b5c6a..c5c9bc25 release-23.04       -> origin/release-23.04
 * [new branch]      feature/merlin-array-dispatch -> origin/feature/merlin-array-dispatch
 * [new branch]      fix-repartition     -> origin/fix-repartition
 * [new branch]      fix-with-properties -> origin/fix-with-properties
 * [new branch]      gh-pages            -> origin/gh-pages
 * [new branch]      laiacano/docs-on-pr -> origin/laiacano/docs-on-pr
 * [new branch]      main                -> origin/main
 * [new branch]      release-22.10       -> origin/release-22.10
 * [new branch]      release-22.11       -> origin/release-22.11
 * [new branch]      release-22.12       -> origin/release-22.12
 * [new branch]      release-23.02       -> origin/release-23.02
 * [new branch]      revert-163-refactor/dictarray-columns -> origin/revert-163-refactor/dictarray-columns
 * [new branch]      stable              -> origin/stable
 * [new branch]      tags-intersection   -> origin

Branch 'main' set up to track remote branch 'main' from 'origin'.
Processing /core
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
  Created wheel for merlin-core: filename=merlin_core-23.5.dev0+28.ge1eaf269-py3-none-any.whl size=164561 sha256=44951aa099780e32b4cd02bc4d0748571a7ba9c437364e634b7da8f0ccc2fba2
  Stored in directory: /tmp/pip-ephem-wheel-cache-1pkiripq/wheels/8f/da/8c/c779661788874afaa32fd10abeac6016635956e3bad9940584
Successfully built merlin-core
  Attempting uninstall: merlin-core
    Found existing installation: merlin-core 23.4.0
    Uninstalling merlin-core-23.4.0:
      Successfully uninstalled merlin-core-23.4.0


From https://github.com/NVIDIA-Merlin/dataloader
 * [new branch]      chore/comprehensive-shapes -> origin/chore/comprehensive-shapes
 * [new branch]      chore/packages-action  -> origin/chore/packages-action
 * [new branch]      collabify_examples     -> origin/collabify_examples
 * [new branch]      docs-add-seo           -> origin/docs-add-seo
 * [new branch]      docs-calver-banner     -> origin/docs-calver-banner
 * [new branch]      ds-api                 -> origin/ds-api
 * [new branch]      embedding_op_with_padding -> origin/embedding_op_with_padding
 * [new branch]      feature/embedding-tags -> origin/feature/embedding-tags
 * [new branch]      fix-sparse-logic       -> origin/fix-sparse-logic
 * [new branch]      fix/tf-batch-size-warning -> origin/fix/tf-batch-size-warning
 * [new branch]      gh-pages               -> origin/gh-pages
 * [new branch]      gha-test               -> origin/gha-test
 * [new branch]      laiacano/docs-pr       -> origin/laiacano/docs-pr
 * [n

Processing /dataloader
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
  Created wheel for merlin-dataloader: filename=merlin_dataloader-23.4.0-py3-none-any.whl size=34732 sha256=907e32a86e091ceba0309393eb6ccaaff913adbe0b56daa93bb1243234deafea
  Stored in directory: /tmp/pip-ephem-wheel-cache-xsaonaut/wheels/8c/19/5b/15dc04f5a977f6a7f73ed66c91996a687b1d9e3154a4765536
Successfully built merlin-dataloader
  Attempting uninstall: merlin-dataloader
    Found existing installation: merlin-dataloader 23.4.0
    Uninstalling merlin-dataloader-23.4.0:
      Successfully uninstalled merlin-dataloader-23.4.0


From https://github.com/NVIDIA-Merlin/NVTabular
   f8f484e5..90489194 release-23.04           -> origin/release-23.04
 * [new branch]      1077-implement          -> origin/1077-implement
 * [new branch]      21.09/column-tagging    -> origin/21.09/column-tagging
 * [new branch]      21.09/dataset-collection -> origin/21.09/dataset-collection
 * [new branch]      21.09/operator-block    -> origin/21.09/operator-block
 * [new branch]      21.09/schema            -> origin/21.09/schema
 * [new branch]      add_sum_to_supported_aggregations -> origin/add_sum_to_supported_aggregations
 * [new branch]      aiobotocore_v2          -> origin/aiobotocore_v2
 * [new branch]      alexanderronquillo-patch-1 -> origin/alexanderronquillo-patch-1
 * [new branch]      atomize_adding_of_tags  -> origin/atomize_adding_of_tags
 * [new branch]      automate_pypi           -> origin/automate_pypi
 * [new branch]      bench-pynvml-fix        -> origin/bench-pynvml-fix
 * [new branch]      branch-0.6       

Branch 'main' set up to track remote branch 'main' from 'origin'.
Processing /nvtabular
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
  Created wheel for nvtabular: filename=nvtabular-23.5.dev0+10.g55fe9344-cp38-cp38-linux_x86_64.whl size=261197 sha256=8146d4d342e5048998f6e578ec2175824deac1a43c36464ffbc7879367000887
  Stored in directory: /tmp/pip-ephem-wheel-cache-h4nasy3b/wheels/df/bf/c2/9cc2a62fe6da42038c26a9c0c4e25f9767093528b102fa30a2
Successfully built nvtabular
  Attempting uninstall: nvtabular
    Found existing installation: nvtabular 23.4.0
    Uninstalling nvtabular-23.4.0:
      Successfully uninstalled nvtabular-23.4.0
Processing /workspace
  Installing build dependencies: started
  Installing build dependencies

From https://github.com/NVIDIA-Merlin/systems
   fce949f..2516efb  release-23.04           -> origin/release-23.04
 * [new branch]      add_xgboost_serving_example -> origin/add_xgboost_serving_example
 * [new branch]      bschifferer-patch-1     -> origin/bschifferer-patch-1
 * [new branch]      bschifferer-patch-2     -> origin/bschifferer-patch-2
 * [new branch]      ci/cpu-action           -> origin/ci/cpu-action
 * [new branch]      dataset-cpu-default-None -> origin/dataset-cpu-default-None
 * [new branch]      docs-nightly-build      -> origin/docs-nightly-build
 * [new branch]      docs-remove-deps        -> origin/docs-remove-deps
 * [new branch]      docs-tox                -> origin/docs-tox
 * [new branch]      docs/contributing       -> origin/docs/contributing
 * [new branch]      docs/coverage-threshold -> origin/docs/coverage-threshold
 * [new branch]      docs/docstring-coverage -> origin/docs/docstring-coverage
 * [new branch]      docs/interrogate-cfg    -> origin/do

Branch 'main' set up to track remote branch 'main' from 'origin'.
Processing /systems
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
  Created wheel for merlin-systems: filename=merlin_systems-23.5.dev0+8.g2b1b90b-py3-none-any.whl size=83188 sha256=e127906aec0187a70c7ef44aea6766a36ffb1ee13479a609a882dcb2cb46f560
  Stored in directory: /tmp/pip-ephem-wheel-cache-atkacac6/wheels/1f/e9/71/1b0c6295aa7f4b37cb70292d96d87d9f38204674e6531bdda6
Successfully built merlin-systems
  Attempting uninstall: merlin-systems
    Found existing installation: merlin-systems 23.4.0
    Uninstalling merlin-systems-23.4.0:
      Successfully uninstalled merlin-systems-23.4.0


From https://github.com/NVIDIA-Merlin/Transformers4Rec
   4a9e7373..911355f4 release-23.04           -> origin/release-23.04
 * [new branch]      DDP_fix                 -> origin/DDP_fix
 * [new branch]      HF-update               -> origin/HF-update
 * [new branch]      add_benchmarking_scripts -> origin/add_benchmarking_scripts
 * [new branch]      add_topk_layer          -> origin/add_topk_layer
 * [new branch]      albert17-check          -> origin/albert17-check
 * [new branch]      batches                 -> origin/batches
 * [new branch]      benfred/datasetschema   -> origin/benfred/datasetschema
 * [new branch]      clean_rnn_block         -> origin/clean_rnn_block
 * [new branch]      core-schema/deprecation-warning -> origin/core-schema/deprecation-warning
 * [new branch]      core-schema/tabular-features -> origin/core-schema/tabular-features
 * [new branch]      core-schema/trainer     -> origin/core-schema/trainer
 * [new branch]      dataloader              -> origin/d

 * [new branch]      v0.1.4-docs             -> origin/v0.1.4-docs
 * [new branch]      v0.1.5-docs             -> origin/v0.1.5-docs
 * [new branch]      v0.1.6-docs             -> origin/v0.1.6-docs
 * [new branch]      v0.1.7-docs             -> origin/v0.1.7-docs
 * [new tag]         v0.1.14                 -> v0.1.14
 * [new tag]         v0.1.15                 -> v0.1.15
 * [new tag]         v0.1.16                 -> v0.1.16
 * [new tag]         v23.02.00               -> v23.02.00
 * [new tag]         v23.05.dev0             -> v23.05.dev0
 * [new tag]           custom_dataloader       -> custom_dataloader
 * [new tag]           v0.1.0                  -> v0.1.0
 * [new tag]           v0.1.1                  -> v0.1.1
 * [new tag]           v0.1.10                 -> v0.1.10
 * [new tag]           v0.1.11                 -> v0.1.11
 * [new tag]           v0.1.12                 -> v0.1.12
 * [new tag]           v0.1.13                 -> v0.1.13
 * [new tag]           v0.1.2   

Branch 'main' set up to track remote branch 'main' from 'origin'.
Processing /transformers4rec
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
    Preparing wheel metadata: started
    Preparing wheel metadata: finished with status 'done'
  Created wheel for transformers4rec: filename=transformers4rec-23.5.dev0+12.g2c9ab409-py3-none-any.whl size=481659 sha256=60160cb11c6d644e8178cdbe9a1f9fc3f4773bdb98599c537aea700b78d657ef
  Stored in directory: /tmp/pip-ephem-wheel-cache-kqpb6x_c/wheels/24/44/e3/c29f7de8e7315585705f880ad32ffeae66fcaeb79003405ef6
Successfully built transformers4rec
  Attempting uninstall: transformers4rec
    Found existing installation: transformers4rec 23.4.0
    Uninstalling transformers4rec-23.4.0:
      Successfully uninstalled transformers4rec-23.4.0


In [2]:
# Copyright 2022 NVIDIA Corporation. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions anda
# limitations under the License.
# ==============================================================================

# Each user is responsible for checking the content of datasets and the
# applicable licenses and determining if suitable for the intended use.

<img src="https://developer.download.nvidia.com/notebooks/dlsw-notebooks/merlin_models_entertainment-with-pretrained-embeddings/nvidia_logo.png" style="width: 90px; float: right;">

# Training with pretrained embeddings

## Overview

In this use case we will consider how we might train with pretrained embeddings.

Pretrained embeddings can allow our model to include information from additional modalities (for instance, we might want to grab CNN descriptors of product images). They can also come from other models that we train on our data. For example, we might train a word2vec model on the sequence of purchased items by a customer and want to include this information in our retrieval or ranking model.

The use cases are many, but this particular example will focus on the technical aspects of working with pretrained embeddings.

We will use a synthetic version of the MovieLens 100k dataset and emulate a scenario where we would have a pretrained embedding for each of the movies in the dataset.

### Learning objectives

- Training with pretrained embeddings
- Understanding [the Schema file](https://github.com/NVIDIA-Merlin/core/blob/main/merlin/schema/schema.py)

## Downloading and preparing the dataset

In [3]:
import merlin.models.tf as mm
from merlin.schema.tags import Tags
import tensorflow as tf
from merlin.models.tf.blocks import *
from merlin.datasets.synthetic import generate_data

import numpy as np

2023-05-16 01:29:04.794838: I tensorflow/core/platform/cpu_feature_guard.cc:194] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
/usr/local/lib/python3.8/dist-packages/merlin/dtypes/mappings/torch.py:43: UserWarning: PyTorch dtype mappings did not load successfully due to an error: No module named 'torch'
  warn(f"PyTorch dtype mappings did not load successfully due to an error: {exc.msg}")


2023-05-16 01:29:07.383103: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-05-16 01:29:07.383461: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-05-16 01:29:07.383586: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero


[INFO]: sparse_operation_kit is imported
[SOK INFO] Import /usr/local/lib/python3.8/dist-packages/merlin_sok-1.1.4-py3.8-linux-x86_64.egg/sparse_operation_kit/lib/libsok_experiment.so
[SOK INFO] Import /usr/local/lib/python3.8/dist-packages/merlin_sok-1.1.4-py3.8-linux-x86_64.egg/sparse_operation_kit/lib/libsok_experiment.so
[SOK INFO] Initialize finished, communication tool: horovod


2023-05-16 01:29:08.980953: I tensorflow/core/platform/cpu_feature_guard.cc:194] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-16 01:29:08.981898: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-05-16 01:29:08.982082: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-05-16 01:29:08.982207: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:998] successful NUMA node read from SysFS had negative value (-1), but there must 

In [4]:
train = generate_data('movielens-100k', num_rows=100_000)
train.schema = train.schema.excluding_by_name(["title"])

In [5]:
target_column = train.schema.select_by_name(["rating_binary"]).column_names[0]
target_column

'rating_binary'

# Passing the embeddings directly to our model using `TensorInitializer`

One way of passing the embeddings directly to our model is using the `TensorInitializer` as part of the `mm.Embeddings`.

This is a straightforward method that works well with small embedding tables.

Let's begin by looking at the schema which holds vital information about our dataset. We can extract the embedding table size for the `moveId` column from it.

In [6]:
train.schema['movieId'].properties['embedding_sizes']['cardinality']

1680.0

From the schema, we can tell that the cardinality of `movieId` is 1680. Index 0 will be used in case an unknown `movieId` is encountered

In order to accommodate this, we initialize our embedding table of dimensionality of (1681, 128).

In [7]:
pretrained_movie_embs = np.random.random((1681, 128))

In [8]:
schema = train.schema

This is only a mock up embedding table. In reality, this is where we would pass our embeddings from another model.

The dimensionality of each embedding, that of 128, is arbitrary. We could have specified some other value here, though generally multiples of 8 tend to work well.

We need to update the schema properties of our `movieId` column since we will not be using the default embedding dimension.

In [9]:
schema['movieId'].properties['embedding_sizes'] = {
    'cardinality': float(pretrained_movie_embs.shape[0]), 
    'dimension': float(pretrained_movie_embs.shape[1])
}

In [10]:
train.schema = schema

In [11]:
train.schema

[{'name': 'movieId', 'tags': {<Tags.ITEM: 'item'>, <Tags.CATEGORICAL: 'categorical'>, <Tags.ID: 'id'>}, 'properties': {'freq_threshold': 0.0, 'max_size': 0.0, 'embedding_sizes': {'cardinality': 1681.0, 'dimension': 128.0}, 'cat_path': './/categories/unique.movieId.parquet', 'num_buckets': None, 'start_index': 0.0, 'domain': {'min': 1, 'max': 1680, 'name': 'movieId'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=(Dimension(min=0, max=None),))), 'is_list': False, 'is_ragged': False}, {'name': 'userId', 'tags': {<Tags.ID: 'id'>, <Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'embedding_sizes': {'cardinality': 943.0, 'dimension': 74.0}, 'num_buckets': None, 'max_size': 0.0, 'start_index': 0.0, 'freq_threshold': 0.0, 'cat_path': './/categories/unique.userId.parquet', 'domain': {'min': 0, 'max': 943, 'name': 'userId'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'genres', 'tags': {<Tags.ITEM: 'item'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'max_size': 0.0, 'start_index': 0.0, 'num_buckets': None, 'freq_threshold': 0.0, 'embedding_sizes': {'cardinality': 216.0, 'dimension': 32.0}, 'cat_path': './/categories/unique.genres.parquet', 'domain': {'min': 0, 'max': 216, 'name': 'genres'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'TE_movieId_rating', 'tags': {<Tags.CONTINUOUS: 'continuous'>}, 'properties': {}, 'dtype': DType(name='float64', element_type=<ElementType.Float: 'float'>, element_size=64, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'userId_count', 'tags': {<Tags.CONTINUOUS: 'continuous'>}, 'properties': {}, 'dtype': DType(name='float32', element_type=<ElementType.Float: 'float'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'gender', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'max_size': 0.0, 'freq_threshold': 0.0, 'embedding_sizes': {'dimension': 16.0, 'cardinality': 2.0}, 'num_buckets': None, 'start_index': 0.0, 'cat_path': './/categories/unique.gender.parquet', 'domain': {'min': 0, 'max': 2, 'name': 'gender'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'zip_code', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'embedding_sizes': {'cardinality': 795.0, 'dimension': 67.0}, 'freq_threshold': 0.0, 'max_size': 0.0, 'cat_path': './/categories/unique.zip_code.parquet', 'start_index': 0.0, 'num_buckets': None, 'domain': {'min': 0, 'max': 795, 'name': 'zip_code'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'rating', 'tags': {<Tags.TARGET: 'target'>, <Tags.REGRESSION: 'regression'>}, 'properties': {}, 'dtype': DType(name='int64', element_type=<ElementType.Int: 'int'>, element_size=64, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'rating_binary', 'tags': {<Tags.BINARY_CLASSIFICATION: 'binary_classification'>, <Tags.TARGET: 'target'>}, 'properties': {}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'age', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'embedding_sizes': {'dimension': 16.0, 'car

## Building the model

We now have everything we need to construct a model and train on our custom embeddings. In order to do so, we will leverage the `TensorInitializer` class and `Embeddings` function to set the `trainable` arg to `False`, so that our pre-trained embedddings will be frozen and not be updated during model training.

In [12]:
embed_dims = {}
embed_dims["movieId"] = pretrained_movie_embs.shape[1]

embeddings_init={
    "movieId": mm.TensorInitializer(pretrained_movie_embs),
}

embeddings_block = mm.Embeddings(
    train.schema.select_by_tag(Tags.CATEGORICAL),
    infer_embedding_sizes=True,
    embeddings_initializer=embeddings_init,
    trainable={'movieId': False},
    dim=embed_dims,
)
input_block = mm.InputBlockV2(train.schema, categorical=embeddings_block)

Let us now feed our input_block into our model.

In [13]:
model = mm.DCNModel(
    train.schema,
    depth=2,
    input_block=input_block,
    deep_block=mm.MLPBlock([64, 32]),
    prediction_tasks=mm.BinaryOutput(target_column)
)

We could have created the model without passing the `embeddings_block` to the `input_block`. The model would still be able to infer how to construct itself (what should be the dimensionality of the input layer and so on) from the information contained in the schema.

However, passing a `TensorInitializer` into the constructor of the `input_block` tells our model to use our embedding table (`pretrained_movie_embs`) for that particular column of our dataset (`movieId`) as opposed to the model randomly initializing a brand new embedding matrix. For categorical columns we do not provide this information, the model will go with the standard initialization logic, which is to create an embedding table of appropriate size and perform random preinitialization.

Additionally, we set the `trainable` parameter for our pre-trained embeddings to `False` to ensure the embeddings will not be modified during training.

## Training

We train our model with `AUC` as our metric.

As we use synthetic data, the AUC score will not improve significantly.

In [14]:
%%time
opt = tf.keras.optimizers.Adagrad(learning_rate=1e-1)
model.compile(optimizer=opt, run_eagerly=False, metrics=[tf.keras.metrics.AUC()])

CPU times: user 13.9 ms, sys: 661 µs, total: 14.6 ms
Wall time: 13.5 ms


In [15]:
model.fit(train, batch_size=1024, epochs=5)

Epoch 1/5


2023-05-16 01:29:20.836517: I tensorflow/compiler/xla/service/service.cc:173] XLA service 0x5aaab040 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2023-05-16 01:29:20.836557: I tensorflow/compiler/xla/service/service.cc:181]   StreamExecutor device (0): Quadro RTX 8000, Compute Capability 7.5
2023-05-16 01:29:20.841893: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:268] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
2023-05-16 01:29:20.975759: I tensorflow/compiler/jit/xla_compilation_cache.cc:480] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


98/98 [==============================] - 12s 73ms/step - loss: 0.6949 - auc: 0.4987 - regularization_loss: 0.0000e+00 - loss_batch: 0.6949
Epoch 2/5
98/98 [==============================] - 2s 21ms/step - loss: 0.6933 - auc: 0.5038 - regularization_loss: 0.0000e+00 - loss_batch: 0.6934
Epoch 3/5
98/98 [==============================] - 1s 11ms/step - loss: 0.6932 - auc: 0.5048 - regularization_loss: 0.0000e+00 - loss_batch: 0.6932
Epoch 4/5
98/98 [==============================] - 1s 9ms/step - loss: 0.6931 - auc: 0.5068 - regularization_loss: 0.0000e+00 - loss_batch: 0.6931
Epoch 5/5
98/98 [==============================] - 1s 9ms/step - loss: 0.6932 - auc: 0.5061 - regularization_loss: 0.0000e+00 - loss_batch: 0.6932


In [16]:
history = model.history.history

# Passing the `EmbeddingOperator` to the `Loader`

Another way of training with pretrained embeddings is to create a custom `Loader` and equip it with the ability to feed embeddings to our model.

When we use `mm.Embeddings` and `TensorInitializer` as above, the embeddings are moved to the GPU and can be considered part of our model. That might become problematic if the embedding table is large.

Taking the approach below the pretrained embeddings are passed to the model as part of each batch. We do not hold the embedding table, which depending on the scenario might consists of millions of rows, in GPU memory.

We can reuse the train data and the embedding information we generated above.

In [17]:
pretrained_movie_embs.shape

(1681, 128)

In [28]:
train.schema = train.schema.without('rating')

In [29]:
from merlin.dataloader.ops.embeddings import EmbeddingOperator

loader = mm.Loader(
    train,
    batch_size=1024,
    transforms=[
        EmbeddingOperator(
            pretrained_movie_embs,
            lookup_key="movieId",
            embedding_name="pretrained_movie_embeddings",
        ),
    ],
)

But we need to recreate the model, as now the  emebeddings will be passed to it straight from the dataloader.

In [30]:
loader.output_schema

[{'name': 'movieId', 'tags': {<Tags.ITEM: 'item'>, <Tags.CATEGORICAL: 'categorical'>, <Tags.ID: 'id'>}, 'properties': {'freq_threshold': 0.0, 'max_size': 0.0, 'embedding_sizes': {'cardinality': 1681.0, 'dimension': 128.0}, 'cat_path': './/categories/unique.movieId.parquet', 'num_buckets': None, 'start_index': 0.0, 'domain': {'min': 1, 'max': 1680, 'name': 'movieId'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=(Dimension(min=0, max=None),))), 'is_list': False, 'is_ragged': False}, {'name': 'userId', 'tags': {<Tags.CATEGORICAL: 'categorical'>, <Tags.USER: 'user'>, <Tags.ID: 'id'>}, 'properties': {'embedding_sizes': {'cardinality': 943.0, 'dimension': 74.0}, 'num_buckets': None, 'max_size': 0.0, 'start_index': 0.0, 'freq_threshold': 0.0, 'cat_path': './/categories/unique.userId.parquet', 'domain': {'min': 0, 'max': 943, 'name': 'userId'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'genres', 'tags': {<Tags.ITEM: 'item'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'max_size': 0.0, 'start_index': 0.0, 'num_buckets': None, 'freq_threshold': 0.0, 'embedding_sizes': {'cardinality': 216.0, 'dimension': 32.0}, 'cat_path': './/categories/unique.genres.parquet', 'domain': {'min': 0, 'max': 216, 'name': 'genres'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'TE_movieId_rating', 'tags': {<Tags.CONTINUOUS: 'continuous'>}, 'properties': {}, 'dtype': DType(name='float64', element_type=<ElementType.Float: 'float'>, element_size=64, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'userId_count', 'tags': {<Tags.CONTINUOUS: 'continuous'>}, 'properties': {}, 'dtype': DType(name='float32', element_type=<ElementType.Float: 'float'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'gender', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'max_size': 0.0, 'freq_threshold': 0.0, 'embedding_sizes': {'dimension': 16.0, 'cardinality': 2.0}, 'num_buckets': None, 'start_index': 0.0, 'cat_path': './/categories/unique.gender.parquet', 'domain': {'min': 0, 'max': 2, 'name': 'gender'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'zip_code', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'embedding_sizes': {'cardinality': 795.0, 'dimension': 67.0}, 'freq_threshold': 0.0, 'max_size': 0.0, 'cat_path': './/categories/unique.zip_code.parquet', 'start_index': 0.0, 'num_buckets': None, 'domain': {'min': 0, 'max': 795, 'name': 'zip_code'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'rating_binary', 'tags': {<Tags.BINARY_CLASSIFICATION: 'binary_classification'>, <Tags.TARGET: 'target'>}, 'properties': {}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element_unit=None, signed=True, shape=Shape(dims=None)), 'is_list': False, 'is_ragged': False}, {'name': 'age', 'tags': {<Tags.USER: 'user'>, <Tags.CATEGORICAL: 'categorical'>}, 'properties': {'embedding_sizes': {'dimension': 16.0, 'cardinality': 8.0}, 'max_size': 0.0, 'start_index': 0.0, 'num_buckets': None, 'cat_path': './/categories/unique.age.parquet', 'freq_threshold': 0.0, 'domain': {'min': 0, 'max': 8, 'name': 'age'}}, 'dtype': DType(name='int32', element_type=<ElementType.Int: 'int'>, element_size=32, element

In [31]:
embeddings_block = mm.Embeddings(
    loader.output_schema.select_by_tag(Tags.CATEGORICAL).remove_col('movieId'),
)

pretrained_embeddings = mm.PretrainedEmbeddings(
    loader.output_schema.select_by_tag(Tags.EMBEDDING)
)

input_block = mm.InputBlockV2(loader.output_schema, categorical=embeddings_block, pretrained_embeddings=pretrained_embeddings)

model = mm.DCNModel(
    loader.output_schema,
    depth=2,
    input_block=input_block,
    deep_block=mm.MLPBlock([64, 32]),
    prediction_tasks=mm.BinaryOutput(target_column)
)

## Training

In [32]:
%%time
opt = tf.keras.optimizers.Adagrad(learning_rate=1e-1)
model.compile(optimizer=opt, run_eagerly=False, metrics=[tf.keras.metrics.AUC()])

model.fit(loader, epochs=5)

KeyError: 'rating_binary'

In [22]:
# history_with_embeddings = model.history.history

The model trains using pretrained embeddings.